### PR TITLE
MultipleFileProperty to support OptionalLoad action

### LINKED ISSUE #17358: MultipleFileProperty to support optional action
`MultipleFileProperty` needs to support `OptionalLoad` action similar to `FileProperty`.

### PR DESCRIPTION
**Pull request by gvardany**
`MultipleFileProperty` does not currently support any actions of `FileProperty`, just defaulting to mandatory `Load` behaviour. This PR implements support for `OptionalLoad` action coherent to `FileProperty`.

**To test:**
1. Pick your favourite algorithm that has `MultipleFileProperty` (e.g. `Load`).
2. Make it optional by adding `action=FileAction.OptionalLoad` (python) or `action=FileProperty::FileAction::OptionalLoad` (C++)\* as a second argument to the `MultipleFileProperty` constructor in property declaration.
3. Make sure that it became optional (e.g. the empty input will now be accepted, while before it would not be)
4. Try what happens with unsupported actions, e.g. `Save`.

*Make sure simpleapi or `FileProperty.h` are imported/included.

Fixes #17358 .

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by samueljackson92** [comment]
Tested this with ```Load``` in C++ and ```IndirectILLReduction``` in Python. Changes work as expected except that ```Load``` still showed the invalid asterisk, despite being an optional property. The Python version worked as expected.

**Comment by gvardany** [comment]
I had a look into this, seems `Load` uses a dedicated dialog, while for other algorithms the dialog is automatically generated, and in this case they behave correctly. Can you try with some other C++ algorithm, like `LoadFITS` ?

**Comment by samueljackson92** [comment]
@gvardany sorry for the slow response on this. I've been away the last few days. I shall try and take a look today.

**Comment by samueljackson92** [comment]
@gvardany This works fine with LoadFITS. :shipit:

File: Framework/API/inc/MantidAPI/MultipleFileProperty.h [hunk 1]
```diff
@@ -119,7 +119,12 @@ namespace API {
 class DLLExport MultipleFileProperty
     : public Kernel::PropertyWithValue<std::vector<std::vector<std::string>>> {
 public:
-  /// Constructor
+  /// Default constructor
```
Comment (SimonHeybrock) [inline Framework/API/inc/MantidAPI/MultipleFileProperty.h]: This is not a default constructor? Also, Doxygen for member methods should be in the source, not in the header (actually you have them in the source, so just remove those from the header).

The same goes for other methods you added to the header file below.

File: Framework/API/src/MultipleFileProperty.cpp [hunk 2]
```diff
@@ -37,17 +37,29 @@ bool doesNotContainWildCard(const std::string &ext) {
...
                                            const std::vector<std::string> &exts)
     : PropertyWithValue<std::vector<std::vector<std::string>>>(
           name, std::vector<std::vector<std::string>>(),
-          boost::make_shared<MultiFileValidator>(exts), Direction::Input),
-      m_multiFileLoadingEnabled(), m_exts(), m_parser(), m_defaultExt("") {
+          boost::make_shared<MultiFileValidator>(
+              exts, (action == FileProperty::Load)),
+          Direction::Input),
+      m_multiFileLoadingEnabled(), m_exts(), m_parser(), m_defaultExt(""),
+      m_action() {
```
Comment (SimonHeybrock) [inline Framework/API/src/MultipleFileProperty.cpp]: No need to explicitly call default constructors (or empty string constructor), can remove those.

File: Framework/API/test/MultipleFilePropertyTest.h [hunk 3]
```diff
@@ -616,6 +618,12 @@ class MultipleFilePropertyTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(fileNames[0].size(), 1);
   }
 
+  void test_multiFileOptionalLoad() {
+    MultipleFileProperty p("Filename", FileProperty::OptionalLoad);
+    p.setValue("myJunkFile.nxs");
+    TS_ASSERT(p.isValid().empty());
+  }
```
Comment (SimonHeybrock) [inline Framework/API/test/MultipleFilePropertyTest.h]: Does it make sense to add a test for the case of no file?

File: Framework/API/inc/MantidAPI/MultipleFileProperty.h (hunk 1 above)
Comment (gvardany) [inline Framework/API/inc/MantidAPI/MultipleFileProperty.h]: Done.

File: Framework/API/src/MultipleFileProperty.cpp (hunk 2 above)
Comment (gvardany) [inline Framework/API/src/MultipleFileProperty.cpp]: Done.

File: Framework/API/test/MultipleFilePropertyTest.h (hunk 3 above)
Comment (gvardany) [inline Framework/API/test/MultipleFilePropertyTest.h]: Added.

**Comment by peterfpeterson** [comment]
@samueljackson92 is this still ready to be merged into master?

**Comment by samueljackson92** [comment]
@peterfpeterson Looks as though @SimonHeybrock's comments have been addressed. This needs the merge conflict to be fixed, but otherwise I'd say :shipit:

**Pull request merged**